File: 275. H-Index II.java
// Time Complexity: O()
// Space Complexity: O()
// Did this code successfully run on Leetcode: YES
// Any problem you faced while coding this: NO

// Your code here along with comments explaining your approach:  Used binary search to find if the index and the value at that index and there after are equal or greater than the index

class Solution {
    public int hIndex(int[] citations) {
        if(citations == null || citations.length == 0) return 0;
        
        int n = citations.length;
        int low = 0, high = n - 1;
        
        while(low <= high){
            int mid = low + (high - low) / 2;
            if(citations[mid] >= n - mid){
                high = mid - 1;
            }
            else{
                low = mid + 1;
            }
        }
        return n - low;
    }
}

File: 275_H_Index_II.py
# Leetcode problem link : https://leetcode.com/problems/h-index-ii/
# Time Complexity:    O(logn) binary search
# Space Complexity:   O(1)
#  Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No

# Your code here along with comments explaining your approach
'''
    Linear approach : BAsed on observation that as soon as the sign flips happen between the number of citations i.e value at an index and its prospective h-index value i.e. n- index then its the h-index value. Linearly calculate the h-index at runtime and return as soon as sign flips
    
    Binary Search : As the array is sorted, same condition can be used to find element using bianry search.

''' 
class Solution:
    def hIndex(self, citations: List[int]) -> int:
        if (citations == None or len(citations) == 0):
            return 0

        n = len(citations)
        lo = 0
        hi = n - 1

        #   iterate till low index and high index cross each other.
        while (lo <= hi):
            mid = lo + int((hi - lo) / 2)
            currCitations = citations[mid]
            # h-index value for mid
            leastCountGreater = n - mid

            #   this will occur when hindex is exactly at this position
            if (currCitations == leastCountGreater):
                return leastCountGreater

            #   if current citations are still less then h-index is in the second half
            elif (currCitations < leastCountGreater):
                lo = mid + 1

            #   this condition means that the location where sign flip happens lies before and so move hi
            else:
                hi = mid - 1

        #   if low and high have crossed then low is the h-index, calculate h-index for index low by length - low index
        leastCountGreater = n - lo
        return leastCountGreater



    def hIndex_linear(self, citations: List[int]) -> int:

        #   edge case
        if (citations == None or len(citations) == 0):
            return 0

        #   initializations
        n = len(citations)

        #   traverse until you find the first flip
        for i in range(n):
            currCitations = citations[i]
            leastCountGreater = n - i

            #   return N-index wherever sign flip happens first
            if (currCitations >= leastCountGreater):
                return leastCountGreater

        return 0